src/librustc/traits/error_reporting.rs

```diff
@@ -272,72 +150,46 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         substs
     }
 
-    fn get_current_failing_impl(&self,
-                                trait_ref: &TraitRef<'tcx>,
-                                obligation: &PredicateObligation<'tcx>)
-                                -> Option<(DefId, subst::Substs<'tcx>)> {
-        let simp = fast_reject::simplify_type(self.tcx,
-                                              trait_ref.self_ty(),
-                                              true);
-        let trait_def = self.tcx.lookup_trait_def(trait_ref.def_id);
-
-        match simp {
-            Some(_) => {
-                let mut matching_impls = Vec::new();
-                trait_def.for_each_impl(self.tcx, |def_id| {
-                    let imp = self.tcx.impl_trait_ref(def_id).unwrap();
-                    let substs = self.impl_substs(def_id, obligation.clone());
-                    let imp = imp.subst(self.tcx, &substs);
-
-                    if self.eq_types(true,
-                                      TypeOrigin::Misc(obligation.cause.span),
-                                      trait_ref.self_ty(),
-                                      imp.self_ty()).is_ok() {
-                        matching_impls.push((def_id, imp.substs.clone()));
-                    }
-                });
-                if matching_impls.len() == 0 {
-                    None
-                } else if matching_impls.len() == 1 {
-                    Some(matching_impls[0].clone())
-                } else {
-                    let end = trait_ref.input_types().len() - 1;
-                    // we need to determine which type is the good one!
-                    Some(matching_impls[get_best_matching_type(&matching_impls,
-                                                               &trait_ref.input_types()[0..end])]
-                                                              .clone())
+    fn impl_with_self_type_of(&self,
+                              trait_ref: ty::PolyTraitRef<'tcx>,
+                              obligation: &PredicateObligation<'tcx>)
+                              -> Option<DefId>
+    {
+        let tcx = self.tcx;
+        let mut result = None;
+        let mut ambiguous = false;
+
+        let trait_self_ty = tcx.erase_late_bound_regions(&trait_ref).self_ty();
+
+        self.tcx.lookup_trait_def(trait_ref.def_id())
+            .for_each_relevant_impl(self.tcx, trait_self_ty, |def_id| {
```

Comment:
so, while I do prefer this simplified approach, it seems like it would mean that `vec[3_i32]` will fail to get the "helpful message" we were shooting for? (I could be confused; I'll kick off a local build and check it out)